crates/wasmparser/src/validator/operators.rs

```diff
@@ -625,6 +634,29 @@ where
         }
     }
 
+    /// Pop a reference type from the operand stack, checking if it matches
+    /// `expected` or the shared version of `expected`. This function will panic
+    /// if `expected` is shared or a concrete type.
+    fn pop_maybe_shared_ref(&mut self, expected: RefType) -> Result<Option<RefType>> {
+        assert!(!self.resources.is_shared_ref_type(expected));
+        let actual = self._calculate_operand_type(Some(expected.into()))?;
+        let actual_ref = match self._as_ref_type(actual)? {
+            Some(rt) => rt,
+            None => return Ok(None),
+        };
+        // Change our expectation based on whether we're dealing with an actual
+        // shared or unshared type.
+        let expected = if self.resources.is_shared_ref_type(actual_ref) {
+            expected
+                .shared()
+                .expect("this only expects abstract heap types")
+        } else {
+            expected
+        };
+        self._check_operand_type(expected.into(), actual)?;
+        Ok(Some(actual_ref))
+    }
```

Comment:
This method feels like roughly the right signature but I'm not sure it requires the other refactorings in this PR. Some changes I might recommend:

* Don't add `is_shared_ref_type` and instead `debug_assert!` here that `expected` is an abstract heap type with the `shared` flag unset
* Use `pop_ref` internally here which doesn't take an `expected` argument
* Given a popped ref test if it's a subtype of the given `expected` either with the `shared` flag set or unset. 
* Return `Result<Option<(RefType, bool)>>` so callers don't have to recalculate `shared`-vs-not.

I think that should satisfy most use cases and avoid the need for larger refactorings?

Reply:
Sure, let's try that. I think we need to keep `is_shared_ref_type` or something like it, though, because we need need to be able to figure out the sharedness of an `actual` concrete type. We can `debug_assert!` away the checks for `expected` but, because `actual` can actually be concrete, we need to be able to resolve the index to the underlying type somehow to inspect its sharedness.

Reply:
A side note: at some point in earlier PRs I had a `RefType::is_shared(&self) -> Option<bool>` that would check sharedness only for the abstract types but I made do without it until now &mdash; even if I re-added it, it wouldn't be enough here because we still have to check sharedness for concrete types.